conbench/entities/history.py

```diff
@@ -85,31 +89,170 @@ def get_history(case_id, context_id, hardware_hash, repo) -> List[tuple]:
+        .over(
+            partition_by=history.c.segment_id,
+            order_by=history.c.timestamp,
+        )
+        .label("rolling_mean_including_this_commit"),
+    ).subquery()
+
+    # Add same column, excluding the current commit
+    history = Session.query(
+        history,
+        func.coalesce(
+            func.avg(history.c.mean).over(
+                partition_by=history.c.segment_id,
+                order_by=s.extract("epoch", history.c.timestamp),
+                range_=(None, -1),
+            ),
+            history.c.mean,
+        ).label("rolling_mean_excluding_this_commit"),
+    ).subquery()
+
+    # Add column with the residuals from the exclusive rolling mean
+    history = Session.query(
+        history,
+        (history.c.mean - history.c.rolling_mean_excluding_this_commit).label(
+            "residual"
+        ),
+    ).subquery()
+
+    # Add column with the rolling standard deviation of the residuals
+    # (these can go outside the segment since we assume they don't change much)
+    history = Session.query(
+        history,
+        func.stddev(history.c.residual)
+        .over(order_by=history.c.timestamp)
+        .label("rolling_sd_including_this_commit"),
+    ).subquery()
+
+    # Add same column, excluding the current commit
+    history = Session.query(
+        history,
+        func.stddev(history.c.residual)
+        .over(
+            order_by=s.extract("epoch", history.c.timestamp),
+            range_=(None, -1),
+        )
+        .label("rolling_sd_excluding_this_commit"),
+    )
+
+    return history.all()
+
+
+def _calc_z_score(
+    data_point: float, unit: str, dist_mean: Optional[float], dist_sd: Optional[float]
+) -> Optional[float]:
+    """Calculate the z-score of a data point compared to a distribution."""
+    if dist_mean is not None and dist_sd is not None and dist_sd != 0:
+        z_score = (data_point - dist_mean) / dist_sd
+    else:
+        z_score = None
+
+    if z_score and _less_is_better(unit):
+        z_score = z_score * -1
+
+    return z_score
+
+
+def append_z_scores(benchmark_results: List[BenchmarkResult]):
+    """Populate the "z_score" attribute of each given BenchmarkResult, comparing it to
+    the distribution of BenchmarkResults that share the same hardware/case/context/repo,
+    ending with the most recent BenchmarkResult in the given BenchmarkResult's commit
+    ancestry on the default branch.
+
+    We choose a commit from the default branch as the end commit, because we probably
+    don't want "bad" commits early on in a PR to set artificially low z-scores later in
+    the PR.
+
+    Should not raise anything. If we can't find a z-score for some reason, the z_score
+    attribute will be None on that BenchmarkResult.
+    """
+    # {(case, context, hardware, repo, commit): (mean, sd)}
+    cached_stats: Dict[
+        Tuple[str, str, str, str, str],
+        Tuple[Optional[float], Optional[float]],
+    ] = {}
```

Comment:
I don't think I saw comments about this above, I'm curious if you could share a bit of how frequently this ends up getting hit | used? My first reaction is that it might be filling in a cache that gets used exactly once for each entry. But I might be missing something!

Reply:
Oh yeah good callout - for most use cases (get one BenchmarkResult, get all BenchmarkResults with one `batch_id`, get all BenchmarkResults with one `run_id`) the cache will only be used once per BenchmarkResult. Not super useful there. But there are a few use cases (get all BenchmarkResults sharing a case name, get the last 1000 BenchmarkResults) where multiple BenchmarkResults could share the same case/context/hardware/repo, and we'd already have the history for the subsequent ones in the cache.

Reply:
Actually, your comment gave me an idea: we could further take advantage of this cache in the `/api/compare` endpoint, by appending all z-scores at once instead of calling `append_z_scores()` on all the baseline BenchmarkResults and then all the contender BenchmarkResults. I'm sure many comparisons share a case/context/hardware/repo. That's sorta the point!